v-next/hardhat-utils/test/bytecode.ts

```diff
@@ -0,0 +1,327 @@
+import type { PrefixedHexString } from "../src/hex.js";
+
+import assert from "node:assert/strict";
+import { describe, it } from "node:test";
+
+import { resolveLinkedBytecode, linkBytecode } from "../src/bytecode.js";
+import { getUnprefixedHexString } from "../src/hex.js";
+
+interface LibraryAddresses {
+  [contractName: string]: PrefixedHexString;
+}
+
+describe("bytecode", () => {
+  describe("resolveLinkedBytecode", () => {
+    describe("validation", () => {
+      it("should throw InvalidLibraryAddressError if a library address is invalid", () => {
+        const artifact = {
+          bytecode:
+            "0xbytecodebytecodebytecode__$placeholderplaceholderplaceholderp$__bytecode",
+          linkReferences: {
+            "contracts/Library.sol": {
+              Library: [{ start: 12, length: 20 }],
+            },
+          },
+        };
+
+        const libraries: LibraryAddresses = {
+          Library: "0xnotanaddress",
+        };
+
+        assert.throws(() => resolveLinkedBytecode(artifact, libraries), {
+          name: "InvalidLibraryAddressError",
+          message: `The following libraries have invalid addresses:
+\t* "Library": "0xnotanaddress"
+
+Please provide valid Ethereum addresses for these libraries.`,
+        });
+      });
+
+      it("should throw AmbiguousLibraryNameError if a library name matches multiple needed libraries", () => {
+        const artifact = {
```

Comment:
Should we add the `Artifact` type explicitly here, to be consistent with `libraries`?

```suggestion
        const artifact: Artifact = {
```

Reply:
Sorry, I should have said that this was a nitpick and that I was happy either way.